query_execution.go

```diff
@@ -822,49 +822,125 @@ func formatResponseDataRec(schema *ast.Schema, selectionSet ast.SelectionSet, re
 //   2. if the target fragments are an implementation of an abstract type, we need to use the __typename from the response body to check which
 //   implementation was resolved. Any fragments that do not match are dropped from the selection set.
-func unionAndTrimSelectionSet(objectTypename string, schema *ast.Schema, selectionSet ast.SelectionSet) ast.SelectionSet {
-	return unionAndTrimSelectionSetRec(objectTypename, schema, selectionSet, map[string]*ast.Field{})
+func unionAndTrimSelectionSet(responseObjectTypeName string, schema *ast.Schema, selectionSet ast.SelectionSet) ast.SelectionSet {
+	filteredSelectionSet := eliminateUnwantedFragments(responseObjectTypeName, schema, selectionSet)
+	return mergeWithTopLevelFragmentFields(filteredSelectionSet)
 }
 
-func unionAndTrimSelectionSetRec(objectTypename string, schema *ast.Schema, selectionSet ast.SelectionSet, seenFields map[string]*ast.Field) ast.SelectionSet {
+func eliminateUnwantedFragments(responseObjectTypeName string, schema *ast.Schema, selectionSet ast.SelectionSet) ast.SelectionSet {
 	var filteredSelectionSet ast.SelectionSet
+
 	for _, selection := range selectionSet {
 		switch selection := selection.(type) {
 		case *ast.Field:
-			if seenField, ok := seenFields[selection.Alias]; ok {
-				if seenField.Name == selection.Name && seenField.SelectionSet != nil && selection.SelectionSet != nil {
-					seenField.SelectionSet = append(seenField.SelectionSet, selection.SelectionSet...)
-				}
-			} else {
-				seenFields[selection.Alias] = selection
-				filteredSelectionSet = append(filteredSelectionSet, selection)
-			}
+			filteredSelectionSet = append(filteredSelectionSet, selection)
+
 		case *ast.InlineFragment:
 			fragment := selection
-			if fragment.ObjectDefinition.IsAbstractType() &&
-				fragmentImplementsAbstractType(schema, fragment.ObjectDefinition.Name, fragment.TypeCondition) &&
-				objectTypenameMatchesDifferentFragment(objectTypename, fragment.TypeCondition) {
+			if shouldEliminateFragment(responseObjectTypeName, schema, fragment.ObjectDefinition, fragment.TypeCondition) {
 				continue
 			}
+			filteredSelectionSet = append(filteredSelectionSet, fragment)
 
-			filteredSelections := unionAndTrimSelectionSetRec(objectTypename, schema, fragment.SelectionSet, seenFields)
-			if len(filteredSelections) > 0 {
-				fragment.SelectionSet = filteredSelections
-				filteredSelectionSet = append(filteredSelectionSet, selection)
-			}
 		case *ast.FragmentSpread:
 			fragment := selection
-			if fragment.ObjectDefinition.IsAbstractType() &&
-				fragmentImplementsAbstractType(schema, fragment.ObjectDefinition.Name, fragment.Definition.TypeCondition) &&
-				objectTypenameMatchesDifferentFragment(objectTypename, fragment.Definition.TypeCondition) {
+			if shouldEliminateFragment(responseObjectTypeName, schema, fragment.ObjectDefinition, fragment.Definition.TypeCondition) {
```

Comment:
Is it possible to unify both these cases? the logic is the same, just the location of the `TypeCondition` on the different structs differs, or does the type system not allow them to be the same?

Reply:
Yeah got about as close as I can without the type system freaking out in the last commit